lib/helpers/macos/applescript.js

```diff
@@ -0,0 +1,117 @@
+/// <reference path="./types.js" />
+
+'use strict';
+
+const { execFile } = require('child_process');
+
+const { KeyboardAction } = require('./KeyboardAction');
+const { KeyCode } = require('./KeyCode');
+const { Modifier } = require('./Modifier');
+const { validateCommand } = require('./validateCommand');
+
+/**
+ * Transform at-driver/webdriver code points into an applescript and run it.
+ * KeyCombination -> AppleKeyCodeCommand -> Applescript -> runScript
+ * @file
+ */
+
+const APPLESCRIPT_RETRY = 3;
+const APPLESCRIPT_TIMEOUT = 5;
+const APPLESCRIPT_TIMEOUT_ERROR = 'AppleEvent timed out';
+
+exports.KeyboardAction = KeyboardAction;
+
+exports.KeyCode = KeyCode;
+
+exports.Modifier = Modifier;
+
+const ScriptKind = (exports.ScriptKind = 'Applescript.Script');
+
+/**
+ * Create an Applescript object.
+ * @param {string} script
+ * @returns {Applescript.Script}
+ */
+const applescript = (exports.applescript = function (script) {
+  return {
+    kind: ScriptKind,
+    script,
+  };
+});
+
+/**
+ * Render a command into an Applescript.
+ * @param {Applescript.Command} command
+ * @returns {Applescript.Script}
+ * @throws
+ */
+exports.renderScript = function (command) {
+  validateCommand(command);
+  const script = `with timeout of ${APPLESCRIPT_TIMEOUT} seconds
+    tell application "System Events"${command.keyCodes
+      .map(
+        code => `
+        key code ${KeyCode[code].toString()}${
+          command.modifiers.length > 1
+            ? ` using {${command.modifiers.map(mod => `${Modifier[mod]} down`).join(', ')}}`
+            : command.modifiers.length === 1
+              ? ` using ${Modifier[command.modifiers[0]]} down`
+              : ``
+        }`,
+      )
```

Comment:
I don't think expressing this as a single Template Literal expression does much for readability. A few variable bindings with expressive names would help folks recognize the code that's being generated.